### ISSUE TITLE
It will die if i transfer to user(who he dont run simplewallet deposit first)(bug)

### ISSUE THREAD

**Issue by ywandy** [askmish/sawtooth-simplewallet#29]
This is my step: 
1. ```sawtooth keygen a```
2. ```sawtooth keygen b```
3. ```simplewallet deposit 100 a```
4. ```simplewallet transfer 25 a b``` 

**(just not deposit any value to b)**

And then,the docker logs will scroll a lot of info cyclically,the info like this: 
```

simplewallet-processor    | Got: transfer and 25
simplewallet-processor    | Got beneficiary: 03605ad5fbb2c9106e71dddb5e1492aaffbf29358cd41b8c02bd061869c873a940
simplewallet-processor    | 479056 [0x7fee26c74780] DEBUG simplewallet null - SimpleWalletApplicator::doTransfer Key: 022116461963790d88e315c0fc81e3f8f2ae74aac5eff42fb888edb69405d34df4 Address: 7e266467df98ca12dc7b4d797a3c13e07718e2066b353a50f73072c4c2f7d3a4fdaaa1
simplewallet-processor    | 479056 [0x7fee26c74780] DEBUG simplewallet null - SimpleWalletApplicator::doTransfer Beneficiary Key: 03605ad5fbb2c9106e71dddb5e1492aaffbf29358cd41b8c02bd061869c873a940 Address: 7e26649edfba4e018ac9c59a0462e8d8565ec4b1db295597da83b7d9819705d090bbe8
simplewallet-processor    | 479057 [0x7fee225c4700] DEBUG sawtooth.MessageDispatcher null - ReceiveMessage MessageType: 8
simplewallet-processor    | 479058 [0x7fee225c4700] DEBUG sawtooth.MessageDispatcher null - ReceiveMessage MessageType: 8
simplewallet-processor    | 479058 [0x7fee26c74780] ERROR sawtooth.TransactionProcessor null - applicator->Apply unknown error
``` 

And if i exec the command ``` simplewallet balance b ``` it will display 404 Not Found.
Like this: 
```
simplewallet balance y
[12:01:14 INFO    connectionpool] Starting new HTTP connection (1): rest-api
[12:01:14 DEBUG   connectionpool] "GET /state/7e26649edfba4e018ac9c59a0462e8d8565ec4b1db295597da83b7d9819705d090bbe8 HTTP/1.1" 404 134
Error: Error 404: Not Found
``` 
I think its solution is:
If i try transfer it will check first if this user had balance(it just can be 0).
Another i think it can initialize user 's balance to 0 if i 'keygen'

Thank you~

**Comment by TheOnlyJoey** [comment]
Had similar issues for a while now (example log https://hastebin.com/asebesevap.php)

We narrowed it down to working with newly created private keys, either trying to transfer tokens from a new account without balance, or to a new account without balance.

The moment we deposit 1 token, everything is fine again.

**Comment by askmish** [comment]
@ywandy @TheOnlyJoey  This behavior is expected and by design. 
The sawtooth keygen just creates a keypair in client systems. Only when you send a deposit txn those keys are used to create an account in sawtooth. The first txn for an account always has to be a deposit txn.

**Comment by TheOnlyJoey** [comment]
But in a situation where you remove deposit (main net, no token minting) you should not be able to do deposit calls anymore, also empty (0) accounts should be supported by just by transfering tokens to a random key right?

Edit: also i guess it should never get into a infinite loop blocking the entire TP?

**Comment by maxmijn** [comment]
@askmish What do you exactly mean by sending a deposit in order to create an account? When I change the python-tp line 153 like this, I can transfer to accounts that have not had a deposit. It also stops the blockchain from crashing when someone that is not in state tries to transfer:

```
if current_entry == []:
  LOGGER.info('No user (debtor) with the key {} '.format(from_key))
  balance = 0
else:
  balance = int(current_entry[0].data)
            
if current_entry_to == []:
  LOGGER.info('No user (creditor) with the key {} '.format(to_key))
  balance_to = 0
else: 
  balance_to = int(current_entry_to[0].data)
```

I'm guessing this is the appropriate behavior, since both the from and to addresses are specified in the TransactionHeader. So the the new address gets added to state when transacting to a new account.

**Comment by maxmijn** [comment]
@askmish @danintel Could you guys have a look at this? I still do not see why a deposit would be needed before a transfer would be possible. Since 1 sawtooth transaction can set multiple addresses in state, the case of not having funds could be handled in the transaction processor logic.

And btw, it looks like the c++ processor goes into a loop every time an InvalidTransaction is thrown. Could this have something to do with the c++ sdk not being supported yet? I do not get this loop when using the python processor.

**Comment by danintel** [comment]
Yes, that is a feature.  The account has to be created before one can transfer or withdraw from the account.  The first deposit also creates the account.  So a workaround is to create a zero-balance account by depositing an amount of 0 before transferring money from another account. 

```
$  sudo docker exec -it simplewallet-client bash
# simplewallet
# sawtooth keygen jack 
# simplewallet deposit 0 jack
# simplewallet balance jack
jack has a net balance of = 0
```

The C++ problem is an unrelated issue--I noticed that too and it is annoying.  My guess is it is because the exception is not properly handled.
Edit: I filed this PR to fix the endless looping with the C++ TP: PR #30
 

For separate, unrelated bugs, I suggest you file separate issues. That will make them easier to track and fix.

**Comment by askmish** [comment]
Closing this issue. A related fix has been raised via PR #30.
